core/common/src/main/java/alluxio/network/connection/ThriftClientPool.java

```diff
@@ -0,0 +1,216 @@
+   */
+  public ThriftClientPool(String serviceName, long serviceVersion, InetSocketAddress address,
+      int maxCapacity, int gcThresholdInSecs) {
+    super(Options.defaultOptions().setMaxCapacity(maxCapacity));
+    mTransportProvider = TransportProvider.Factory.create();
+    mServiceName = serviceName;
+    mServiceVersion = serviceVersion;
+    mAddress = address;
+    mGcThresholdInSecs = gcThresholdInSecs;
+  }
+
+  @Override
+  protected void closeResource(T client) {
+    // Note that the input and output protocol is the same in Alluxio.
+    TTransport transport = client.getOutputProtocol().getTransport();
+    if (transport.isOpen()) {
+      LOG.info("Closing thrift transport.");
+      transport.close();
+    } else {
+      LOG.info("Close a closed thrift transport.");
+    }
+  }
+
+  @Override
+  protected void closeResourceSync(T client) {
+    closeResource(client);
+  }
+
+  /**
+   * Creates a thrift client instance.
+   *
+   * @return the thrift client created
+   * @throws IOException if it fails to create a thrift client
+   */
+  @Override
+  protected T createNewResource() throws IOException {
+    TTransport transport = mTransportProvider.getClientTransport(mAddress);
+    TProtocol binaryProtocol = new TBinaryProtocol(transport);
+    T client = createThriftClient(new TMultiplexedProtocol(binaryProtocol, mServiceName));
+
+    RetryPolicy retry =
+        new ExponentialBackoffRetry(CONNECTION_OPEN_RETRY_BASE_SLEEP_MS, Constants.SECOND_MS,
+            CONNECTION_OPEN_RETRY_MAX);
+    while (true) {
+      try {
+        if (!transport.isOpen()) {
+          transport.open();
+        }
+        if (transport.isOpen()) {
+          checkVersion(client);
+        }
+      } catch (TTransportException e) {
+        LOG.error(
+            "Failed to connect (" + retry.getRetryCount() + ") to " + getServiceNameForLogging()
+                + " @ " + mAddress, e);
+        if (!retry.attemptRetry()) {
+          throw new IOException(e);
+        }
+      }
+      break;
+    }
+
+    return client;
+  }
+
+  /**
+   * Checks whether a client is healthy.
+   *
+   * @param client the thrift client to check
+   * @return true if the client is open (i.e. connected)
+   */
+  @Override
+  protected boolean isHealthy(T client) {
+    return client.getOutputProtocol().getTransport().isOpen();
+  }
+
+  @Override
+  protected boolean shouldGc(ResourceInternal<T> clientResourceInternal) {
+    return System.currentTimeMillis() - clientResourceInternal.getLastAccessTimeMs()
+        > (long) mGcThresholdInSecs * (long) Constants.SECOND_MS;
+  }
+
+  /**
+   * Check the service version to see whether it matches the expected version.
+   *
+   * @param client the client
+   * @throws IOException if it fails to check version
+   */
+  private void checkVersion(T client) throws TTransportException, IOException {
+    if (mIsVersionChecked) {
+      return;
+    }
```

Comment:
yes, it does fail fast.  It is slightly cleaner not to cache the version because 
 it is possible that we cannot even get the version because version mismatching which will be cached as well (not a big deal though) in the current implementation.

Do you prefer to change it?